torch/optim/adam.py

```diff
@@ -47,6 +47,20 @@ def __init__(self, params, lr=1e-3, betas=(0.9, 0.999), eps=1e-8,
                         weight_decay=weight_decay, amsgrad=amsgrad)
         super(Adam, self).__init__(params, defaults)
 
+        # State initialization
+        for group in self.param_groups:
+            for p in group['params']:
+                state = self.state[p]
+                state['step'] = 0
+                # Exponential moving average of gradient values
+                state['exp_avg'] = torch.zeros_like(p, memory_format=torch.preserve_format)
```

Comment:
I don't think we want to move these to the init.
Some new features like [this](https://github.com/pytorch/pytorch/pull/44538#issuecomment-697170994) actually relies on the fact that our optimizers lazily initialize their buffers.